### ISSUE TITLE
davical and debian 10 not working

### ISSUE THREAD

**Issue by Mondhund** [python-caldav/caldav#130]
Hi,

(using caldav-0.7,1)
after a server upgrade from debian 8, davical 1.1.3.1-1+deb8u1, php 5.6.40 to debain 10, davical 1.1.8-1+deb10u1, php 7.3 
a call to principal.calendars() only returns an empty string. 

Using curl for the same calendar gives the correct output. It seems the difference in call is in the header. With curl it's:
`
curl --data '<?xml version="1.0" encoding="utf-8" ?><propfind xmlns="DAV:"><prop><displayname/><resourcetype/></prop></propfind>' --header 'Content-Type: text/xml' --header 'Depth: 2' -X PROPFIND -s -u 'someuser:somepass' https://<url of calendar>`

And caldav debug output shows:
`
DEBUG:caldav:sending request - method=PROPFIND, url=https://<server>:443/<path>, headers={'User-Agent': 'Mozilla/5.0', 'Content-Type': 'text/xml', 'Accept': 'text/xml, text/calendar', 'Depth': '0'}
`
The difference seems to be the 'Depth':'2'.  Any ideas how to fix this?

**Comment by tobixen** [comment]
Currently I only have a very old version of DAViCal to test against, and I don't have capacity to do anything with that currently.  By any chance, is it possible for you to set up a test account on your server so I can run the test suite towards it?

**Comment by tobixen** [comment]
`'Depth': '0'` should work, as far as I can understand

**Comment by Mondhund** [comment]
> `'Depth': '0'` should work, as far as I can understand

Yes, but if I replace in the curl-call 'Depth: 2' with 'Depth: 0' I also get an empty list just like with the python call:
```
 curl --data '<?xml version="1.0" encoding="utf-8" ?><propfind xmlns="DAV:"><prop><displayname/><resourcetype/></prop></propfind>' --header 'Content-Type: text/xml' --header 'Depth: 0' -X PROPFIND -s -u 'someuser:somepass' https://<server>
<?xml version="1.0" encoding="utf-8" ?>
<multistatus xmlns="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
 <response>
  <href> ... </href>
  <propstat>
   <prop>
    <displayname> ... </displayname>
    <resourcetype>
     <collection/>
     <C:calendar/>
    </resourcetype>
   </prop>
   <status>HTTP/1.1 200 OK</status>
  </propstat>
 </response>
</multistatus>
```
Disclaimer: from my side it's wild guessing since I'm not deep into this topic :) 

Wrt. the test-server, we're about to set something up. Should be available in the next hour or so.

Thanks ...

**Comment by tobixen** [comment]
While you're at it, can you check what happens if you specify "Depth: 1"?

Perhaps I should read up on that depth-header, but as far as I remember "depth: 0" should return everything.  It's possible to change the default value to 2.  I don't have enough details to deduct where this should be done, but probably in objects.py, method calendar_home_set, the get_properties there accepts a depth parameter.

I would have to test it towards multiple calendar server implementations as possible before doing such a change in the master branch or in the next release.

**Comment by Mondhund** [comment]
> While you're at it, can you check what happens if you specify "Depth: 1"?

this gives via curl the same output as 'Depth: 2'

**Comment by tobixen** [comment]
And can you check what happens if you give the depth-parameter to the get_properties in the calendar_home_set?

-- 
Tobias Brox
Senior System Consultant
Redpill Linpro AS - Changing the game
Mobil: +47 917 000 50 
Kontor: +47 215 441 68

**Comment by Mondhund** [comment]
then I'll get in the debug a long list of *.ics entries

`DEBUG:caldav:raw response: b'<?xml version="1.0" encoding="utf-8" ?>\n<multistatus xmlns="DAV:">\n <response>\n  <href> ...</href>\n  <propstat>\n   <prop>\n    <current-user-principal>\n     <href>...</href>\n    </current-user-principal>\n   </prop>\n   <status>HTTP/1.1 200 OK</status>\n  </propstat>\n </response>\n <response>\n  <href>.../calendar_entry.ics</href>\n
(long list of output...)`

but the  final call still returns an empty string.

```
#!/usr/bin/python3

import logging
import caldav
from caldav.elements import dav, cdav
from pprint import pprint
logging.basicConfig(level=logging.DEBUG)
url = "https://<url>"
client = caldav.DAVClient(url)
principal = client.principal()
calendars = principal.calendars()
pprint (calendars)
```
returns []

Here's the full debug output with the original depth=0 (urls, usenames removed):

```
./test-caldav.py
DEBUG:caldav:url: https://user:user@url/pwdavical/caldav.php/path/
DEBUG:caldav:self.url: https://user:user@url/pwdavical/caldav.php/path/
DEBUG:caldav:sending request - method=PROPFIND, url=https://url:443/pwdavical/caldav.php/path/, headers={'User-Agent': 'Mozilla/5.0', 'Content-Type': 'text/xml', 'Accept': 'text/xml, text/calendar', 'Depth': '0'}
body:
<?xml version='1.0' encoding='utf-8'?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav"><D:prop><D:current-user-principal/></D:prop></D:propfind>
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): url:443
DEBUG:urllib3.connectionpool:https://url:443 "PROPFIND /pwdavical/caldav.php/path/ HTTP/1.1" 401 40
DEBUG:caldav:response headers: {'Date': 'Mon, 18 Jan 2021 14:11:32 GMT', 'Server': 'Apache', 'X-Frame-Options': 'sameorigin', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'WWW-Authenticate': 'Basic realm="Company Calendar Server"', 'Content-Length': '40', 'Keep-Alive': 'timeout=5, max=100', 'Connection': 'Keep-Alive', 'Content-Type': 'text/plain; ; charset="utf-8"'}
DEBUG:caldav:response status: 401
DEBUG:caldav:raw response: b'Please log in for access to this system.'
DEBUG:urllib3.connectionpool:https://url:443 "PROPFIND /pwdavical/caldav.php/path/ HTTP/1.1" 207 350
DEBUG:caldav:response headers: {'Date': 'Mon, 18 Jan 2021 14:11:32 GMT', 'Server': 'Apache', 'X-Frame-Options': 'sameorigin', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'DAV': '1, 2, 3, access-control, calendar-access, calendar-schedule, extended-mkcol, bind, addressbook, calendar-proxy', 'ETag': '"3f98f3736ec2c3a9f9dc6345826329d7"', 'X-DAViCal-Version': 'DAViCal/1.1.8; DB/1.3.3', 'Content-Length': '350', 'Keep-Alive': 'timeout=5, max=99', 'Connection': 'Keep-Alive', 'Content-Type': 'text/xml; charset="utf-8"'}
DEBUG:caldav:response status: 207
DEBUG:caldav:raw response: b'<?xml version="1.0" encoding="utf-8" ?>\n<multistatus xmlns="DAV:">\n <response>\n  <href>/pwdavical/caldav.php/path/</href>\n  <propstat>\n   <prop>\n    <current-user-principal>\n     <href>/pwdavical/caldav.php/user/</href>\n    </current-user-principal>\n   </prop>\n   <status>HTTP/1.1 200 OK</status>\n  </propstat>\n </response>\n</multistatus>\n'
DEBUG:caldav:sending request - method=PROPFIND, url=https://url:443/pwdavical/caldav.php/user/, headers={'User-Agent': 'Mozilla/5.0', 'Content-Type': 'text/xml', 'Accept': 'text/xml, text/calendar', 'Depth': '0'}
body:
<?xml version='1.0' encoding='utf-8'?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav"><D:prop><C:calendar-home-set/></D:prop></D:propfind>
DEBUG:urllib3.connectionpool:https://url:443 "PROPFIND /pwdavical/caldav.php/user/ HTTP/1.1" 207 375
DEBUG:caldav:response headers: {'Date': 'Mon, 18 Jan 2021 14:11:32 GMT', 'Server': 'Apache', 'X-Frame-Options': 'sameorigin', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'DAV': '1, 2, 3, access-control, calendar-access, calendar-schedule, extended-mkcol, bind, addressbook, calendar-proxy', 'ETag': '"464b14a187da0603db734b58cd6f2019"', 'X-DAViCal-Version': 'DAViCal/1.1.8; DB/1.3.3', 'Content-Length': '375', 'Keep-Alive': 'timeout=5, max=98', 'Connection': 'Keep-Alive', 'Content-Type': 'text/xml; charset="utf-8"'}
DEBUG:caldav:response status: 207
DEBUG:caldav:raw response: b'<?xml version="1.0" encoding="utf-8" ?>\n<multistatus xmlns="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">\n <response>\n  <href>/pwdavical/caldav.php/user/</href>\n  <propstat>\n   <prop>\n    <C:calendar-home-set>\n     <href>/pwdavical/caldav.php/user/</href>\n    </C:calendar-home-set>\n   </prop>\n   <status>HTTP/1.1 200 OK</status>\n  </propstat>\n </response>\n</multistatus>\n'
DEBUG:caldav:sending request - method=PROPFIND, url=https://url:443/pwdavical/caldav.php/user/, headers={'User-Agent': 'Mozilla/5.0', 'Content-Type': 'text/xml', 'Accept': 'text/xml, text/calendar', 'Depth': '1'}
body:
<?xml version='1.0' encoding='utf-8'?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav"><D:prop><D:resourcetype/><D:displayname/></D:prop></D:propfind>
DEBUG:urllib3.connectionpool:https://url:443 "PROPFIND /pwdavical/caldav.php/user/ HTTP/1.1" 207 1087
DEBUG:caldav:response headers: {'Date': 'Mon, 18 Jan 2021 14:11:32 GMT', 'Server': 'Apache', 'X-Frame-Options': 'sameorigin', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'DAV': '1, 2, 3, access-control, calendar-access, calendar-schedule, extended-mkcol, bind, addressbook, calendar-proxy', 'ETag': '"65b04f66e14747656f18fb08e1dfa39a"', 'X-DAViCal-Version': 'DAViCal/1.1.8; DB/1.3.3', 'Content-Length': '1087', 'Keep-Alive': 'timeout=5, max=97', 'Connection': 'Keep-Alive', 'Content-Type': 'text/xml; charset="utf-8"'}
DEBUG:caldav:response status: 207
DEBUG:caldav:raw response: b'<?xml version="1.0" encoding="utf-8" ?>\n<multistatus xmlns="DAV:" xmlns:C="http://calendarserver.org/ns/">\n <response>\n  <href>/pwdavical/caldav.php/user/</href>\n  <propstat>\n   <prop>\n    <resourcetype>\n     <collection/>\n    
 <principal/>\n    </resourcetype>\n    <displayname>NOC-Rufbereitschafts-User</displayname>\n   </prop>\n   <status>HTTP/1.1 200 OK</status>\n  </propstat>\n </response>\n <response>\n  <href>/pwdavical/caldav.php/user/calendar-proxy-read/</href>\n  <propstat>\n 
  <prop>\n    <resourcetype>\n     <principal/>\n     <collection/>\n     <C:calendar-proxy-read/>\n    </resourcetype>\n    <displayname>/user/calendar-proxy-read/</displayname>\n   </prop>\n   <status>HTTP/1.1 200 OK</status>\n  </propstat>\n </response>\n <response>\n  <href>/pwdavical/caldav.php/user/calendar-proxy-write/</href>\n  <propstat>\n   <prop>\n    <resourcetype>\n     <principal/>\n     <collection/>\n     <C:calendar-proxy-write/>\n    </resourcetype>\n    <displayname>/user/calendar-proxy-write/</displayname>\n   </prop>\n   <status>HTTP/1.1 200 OK</status>\n  </propstat>\n </response>\n</multistatus>\n'
[]
```

**Comment by tobixen** [comment]
I said that "Depth: 0" should return everything, that was obviously a brainfart.  Here is something from the WebDAV RFC:

>  A client MUST submit a Depth header with a value of "0", "1", or "infinity" with a PROPFIND request.  Servers MUST support "0" and "1"  depth requests on WebDAV-compliant resources and SHOULD support "infinity" requests.  In practice, support for infinite-depth requests MAY be disabled, due to the performance and security concerns associated with this behavior.  Servers SHOULD treat a request without a Depth header as if a "Depth: infinity" header was included.

It doesn't say much explicitly on exactly how the depth header should be interpreted in the PROPFIND scope.  Perhaps the CalDAV standard outlines more on that, I didn't check it up.  If I understand it correctly, when yielding "depth: 0", the properties of the object requested should be given, and when yielding "depth: 1" the properties for the direct children should also be given.  Apparently DAViCal interprets "Depth: 2" when requesting the "calendar home set" so that it should return all calendars (direct children of the calendar home set) and events (grandchildren).

The initial URL for DAViCal should be https://url:443/pwdavical/caldav.php/ - not https://url:443/pwdavical/caldav.php/path/ - does that resolve the problem?

**Comment by Mondhund** [comment]
I already tried shortening the pathname, but the result is always an empty string.

**Comment by tobixen** [comment]
> ```
> #!/usr/bin/python3
> 
>(...)
> principal = client.principal()
> calendars = principal.calendars()

either the new version of DAViCal is doing something wrong and everyone else is doing something right, or the new version of DAViCal is doing something right and everyone else is doing it wrong, or perhaps the standard is unclear on something.

try something like this (not tested, so I've possibly slipped in some bugs below):

```python
def print_url(obj):
    print(obj.url.path.replace('yourusername', '$user))

principal = client.principal()
print_url(principal)
calendar_home_set = principal.calendar_home_set
print_url(calendar_home_set)
calendars = calendar_home_set.calendars()
pprint(calendars)
```

Most likely the problem is in the last step here (`calendars = calendar_home_set.calendars()`), but I want to be sure on that before spending more time on it.  It's still a bit unclear for me exactly which request is failing.  In the top you write about some query going towards a calendar URL, but I don't think the communication above does any propfind on the individual calendars found (and also, a propfind with depth 1 on the calendars should by the WebDAV definition yield calendar events ... that's not something we want to fetch from the server at this point).

**Comment by Mondhund** [comment]
output is always:

/pwdavical/caldav.php/$user/
/pwdavical/caldav.php/$user/

regardless if I use the short version of the url or not or if I set depth=0 or 1 ind objects.py

**Comment by tobixen** [comment]
Interesting ... and it may be right, the URL for the calendar home set may be the same as the URL for the principal in DAViCal.  Or it may be a bug.  Can you add a debug hook in objects.py, there is a comment there like this:

     ## TODO: sanitized_url should never be None

can you check if sanitized_url is None there?

In the debug output above you write `/pwdavical/caldav.php/user/`, I thought it was literal 'user', but it's probably sth like  `/pwdavical/caldav.php/Mondhund/`?

**Comment by tobixen** [comment]
Here is the bottom of your debug logging, rewritten for clarity and with some comments:

```
DEBUG:caldav:sending request
PROPFIND https://url:443/pwdavical/caldav.php/user/

  User-Agent: Mozilla/5.0
  Content-Type: text/xml
  Accept: text/xml, text/calendar
  Depth: 1

  <?xml version='1.0' encoding='utf-8'?>
  <D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
      <D:prop>
          <D:resourcetype/>
          <D:displayname/>
     </D:prop>
  </D:propfind>
```

This one is supposed to fetch all the calendars, with displayname.  Depth is 1.  It may be that the URL it gets is wrong, so that it's trying to get the user with all calendar sets rather than all calendars within a calendar set.  In that case changing to Depth=2 may help.

```
DEBUG:caldav:response headers:
Date: Mon, 18 Jan 2021 14:11:32 GMT
Server: Apache
X-Frame-Options: sameorigin
Strict-Transport-Security: max-age=31536000; includeSubDomains
DAV: 1, 2, 3, access-control, calendar-access, calendar-schedule, extended-mkcol, bind, addressbook, calendar-proxy
ETag: '"65b04f66e14747656f18fb08e1dfa39a"
X-DAViCal-Version: 'DAViCal/1.1.8; DB/1.3.3
Content-Length: 1087
Keep-Alive: timeout=5, max=97
Connection: Keep-Alive
Content-Type: text/xml; charset="utf-8"

DEBUG:caldav:response status: 207

DEBUG:caldav:raw response: b'
<?xml version="1.0" encoding="utf-8" ?>
<multistatus xmlns="DAV:" xmlns:C="http://calendarserver.org/ns/">
 <response>
  <href>/pwdavical/caldav.php/user/</href>
  <propstat>
   <prop>
    <resourcetype>
     <collection/>
     <principal/>
    </resourcetype>
    <displayname>Some User</displayname>
   </prop>
   <status>HTTP/1.1 200 OK</status>
  </propstat>
 </response>
 <response>
  <href>/pwdavical/caldav.php/user/calendar-proxy-read/</href>
  <propstat>
   <prop>
    <resourcetype>
     <principal/>
     <collection/>
     <C:calendar-proxy-read/>
    </resourcetype>
    <displayname>/user/calendar-proxy-read/</displayname>
   </prop>
   <status>HTTP/1.1 200 OK</status>
  </propstat>
 </response>
 <response>
  <href>/pwdavical/caldav.php/user/calendar-proxy-write/</href>
  <propstat>
  <prop>
    <resourcetype>
     <principal/>
     <collection/>
     <C:calendar-proxy-write/>
    </resourcetype>
    <displayname>/user/calendar-proxy-write/</displayname>
   </prop>
   <status>HTTP/1.1 200 OK</status>
 </propstat>
 </response>
</multistatus>
[]
```

**Comment by tobixen** [comment]
So it seems like a "principal" and a "calendar set" is the same (since resourcetype is set both to collection and calendar-proxy-write).  It also explicitly states a bit above that the calendarhomeset is <href>/pwdavical/caldav.php/user/</href>.

I find it strange, and it seems a bit like it's a bug with davical.  We're asking for the "calendarhomeset" of (perhaps) `/pwdavical/caldav.php/Mondhund` and we're getting back ...  `/pwdavical/caldav.php/Mondhund`.  Then we ask for the children of `/pwdavical/caldav.php/Mondhund/`, and we get back ... `/pwdavical/caldav.php/Mondhund/` (plus some strange proxy URLs).

A very ugly workaround may be to change this:

```
    def children(self, type=None):
        """
        List children, using a propfind (resourcetype) on the parent object,
        at depth = 1.
        """
        c = []

        depth = 1
```

into

```
    def children(self, type=None):
        """
        List children, using a propfind (resourcetype) on the parent object,
        at depth = 1.
        """
        c = []

        depth = 2
```

that may perhaps work in your case, but it's not something I would put into the code, for obvious reasons.

**Comment by tobixen** [comment]
I would suggest to ask the DAViCal developers for help, point them to this issue eventually.

**Comment by Mondhund** [comment]
> Interesting ... and it may be right, the URL for the calendar home set may be the same as the URL for the principal in DAViCal. Or it may be a bug. Can you add a debug hook in objects.py, there is a comment there like this:
> 
> ```
>  ## TODO: sanitized_url should never be None
> ```
> 
> can you check if sanitized_url is None there?
> 
> In the debug output above you write `/pwdavical/caldav.php/user/`, I thought it was literal 'user', but it's probably sth like `/pwdavical/caldav.php/Mondhund/`?

it's not None, but /pwdavical/caldav.php/$user.  Btw: the original path is /pwdavical/caldav.php/$company_name/$user. Not sure if that's a point?

**Comment by tobixen** [comment]
I don't think that matters much (though, it may be interesting to know what kind of resourcetype is found at `/pwdavical/caldav.php/$company_name`)

**Comment by Mondhund** [comment]
The test-server is now ready. You can access ist with pycaldav:pycaldav at https://cal-dev.dfn.de/pwdavical/caldav.php/dfn/noc.events. 

It has a few test entries that you should be able to access:

```
 curl --data '<?xml version="1.0" encoding="utf-8" ?><propfind xmlns="DAV:"><prop><displayname/><resourcetype/></prop></propfind>' --header 'Content-Type: text/xml' --header 'Depth: 2' -X PROPFIND -s -u 'pycaldav:pycaldav' https://cal-dev.dfn.de/pwdavical/caldav.php/dfn/noc.events
<?xml version="1.0" encoding="utf-8" ?>
<multistatus xmlns="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
 <response>
  <href>/pwdavical/caldav.php/dfn/noc.events/</href>
  <propstat>
   <prop>
    <displayname>NOC Events</displayname>
    <resourcetype>
     <collection/>
     <C:calendar/>
    </resourcetype>
   </prop>
   <status>HTTP/1.1 200 OK</status>
  </propstat>
 </response>
 <response>
  <href>/pwdavical/caldav.php/dfn/noc.events/ac6c66b3-405f-4f2d-9cd9-5252b810ba83.ics</href>
  <propstat>
   <prop>
    <displayname>testing1</displayname>
    <resourcetype/>
   </prop>
   <status>HTTP/1.1 200 OK</status>
  </propstat>
 </response>
 <response>
  <href>/pwdavical/caldav.php/dfn/noc.events/055d95bf-4485-4edd-a11d-0bb340652f7b.ics</href>
  <propstat>
   <prop>
    <displayname>testing1</displayname>
    <resourcetype/>
   </prop>
   <status>HTTP/1.1 200 OK</status>
  </propstat>
 </response>
</multistatus>
```

**Comment by Mondhund** [comment]
Hi,
> Currently I only have a very old version of DAViCal to test against, and I don't have capacity to do anything with that currently. By any chance, is it possible for you to set up a test account on your server so I can run the test suite towards it?

not sure if you had time to check the test installation. Is the access working?
Thanks!

**Comment by tobixen** [comment]
I didn't get time yesterday.  I'll have a quick look at it now.

**Comment by tobixen** [comment]
I did run some functional tests towards it, observed one 500 internal server error.  Interesting.  I'll look more into it after lunch, if I get time.

**Comment by tobixen** [comment]
Hm.

I used https://cal-dev.dfn.de/pwdavical/caldav.php/ as the starting URL and ran tests, disabling tests that didn't work, here is what I found:

* journal entries seems not to be supported (that's whats giving 500 internal server error.  Well, who uses journal entries anyway ...)
* vtodos without a dtstart does not show up when doing date search on vtodo's.  Not sure if that is a bug or not.
* In the sync test, at one point, the server unexpectedly returned two objects instead of one.  This is not a breach of the standard, but a bit annoying when doing testing.

Earlier versions of davical did not support freebusy requests, this seems to work now.

I will look more into the issues a bit later.

**Comment by tobixen** [comment]
TLDR: I can't reproduce any problems.

The calendar `dfn/noc.events` does not show up when I do
I can do this without problems:

```python
my_client = caldav.DAVClient(
    url='https://cal-dev.dfn.de/pwdavical/caldav.php/', 
    username='pycaldav', 
    password='pycaldav'
calendars = my_client.principal().calendars()
print(calendars)
```

No errors raised, but the calendar you linked to is not included in the results.

The calendar can be accessed by using the Calendar constructor and specifying the URL.  The URL may be relative compared to the URL given to the client object.

```python
    mycal = caldav.Calendar(client=client, url='dfn/noc.events/')
    print(mycal.events())
```

**Comment by Mondhund** [comment]
thanks for your research - I'm busy today so I'll check later. First test indeed gives me a calendar back. Since it's been working before for years and we have been happily entering and deleting *ics in the calendars, It does looks like something in the semantic on the server side has changed. I'll dig deeper into this later.

**Comment by tobixen** [comment]
Perhaps something have gone wrong while doing data migration.  Apparently you have a problem with your calendars or users, while I see no problems with the test user.

**Comment by Mondhund** [comment]
thanks - made it work using your code. Not sure why the old version was no longer working. It seems the server returned in the old release via the principal a list of calendars under the url that we iterated through to find the correct calendar events and now in the new version, it returns an empty string unless you use the caldav.Calendar call with the path as you suggested. Perhaps the old method should never have worked. Anyhow - thanks again and great work!
